### DISCUSSION TITLE
Introducing Amazing Oqtane Studio

### DISCUSSION THREAD

**Discussion by vnetonline** [oqtane/oqtane.framework#4199] [Polls]
[ Amazing Oqtane Studi](https://github.com/AmazingSoftwareSolutions/Amazing.Oqtane.Studio/)o helps you generate code that conforms to the Oqtane Framework for your Modules

- Create Entities
- Add Properties to the Entities
- Add your Navigation Collections (many to many) and Properties (1 to many)

Then let Amazing Oqtane Studio Generate
- Models
- EntityBuilders for your Models
- CRUD Server Services
- Repositories for your Entities
- Managers
- Controllers
- Build your UI (Index, Edit, Add)

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/54ee2bc5-531c-48fc-b54f-dc598f36b8ad)

**Comment by sbwalker** [comment]
@vnetonline are you referring to a CLI which is able to scaffold specific classes rather than scaffolding an entire project (which is what the framework's template capability does currently)?

**Reply by vnetonline** [reply]
Yes that's right the Amazing Oqtane Studio will be able to generate specify classes like build the Model, Repository, EntityBuilder, ServerEntityServices etc etc based on Entity and Properties you supply via the UI. CLI will be phase two of this project

**Comment by vnetonline** [comment]
I have made some progress on the Amazing Oqtane Studio project; generating Models with Properties is now up and running.

![screencast_amazing-oqtane-studio](https://github.com/oqtane/oqtane.framework/assets/15242923/cd5c0d26-6643-4370-b61c-d4ed1f95afed)

**Comment by vnetonline** [comment]
Good progress has been made on the auto-generation of the EntityBuilderTemplate 

![screencast-amazing-oqtane-studio-EntityBuilderTemplate](https://github.com/oqtane/oqtane.framework/assets/15242923/9abcf333-7e9c-4c69-84f8-b6c8bef6b4b5)

**Comment by vnetonline** [comment]
Auto-generation of the EntityBuilderTemplate with all Oqtane types complete

![screencast-amazing-oqtane-studio-EntityBuilderTemplate-Complete](https://github.com/oqtane/oqtane.framework/assets/15242923/885c1913-3dcc-4581-8fef-2ed36c9ba7a7)

**Comment by mdmontesinos** [comment]
Is there any publicly available demo for this product?

**Comment by vnetonline** [comment]
I just created a dedicated page on https://LearnOqtane.com 

check it out

https://learnoqtane.com/modules/amazing-oqtane-studio

Register your interest button coming soon.

**Comment by vnetonline** [comment]
Amazing Oqtane Studio is now able to generate a Repository Interface for an entity

![screencast-amazing-oqtane-studio-model-interface-entitybuilder](https://github.com/oqtane/oqtane.framework/assets/15242923/6e04e9a1-586a-4874-9ba4-aee6fa1ac2d5)

**Comment by sbwalker** [comment]
@vnetonline this is looking really good! The one question I have is how you plan to support versioning? The standard module templates  have a template.json which contains a Version property. This property indicates the minimum version which the template supports. For example, the current Default Module Template has a version of 5.1.0 which means that the template can only be used with Oqtane Framework versions 5.1.0 and higher:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/6d851dc2-1041-47c3-a15a-b72367dcaa53)

Basically, anytime a module template is updated to leverage new framework features, the template.json needs to be updated. Traditionally this has occurred fairly frequently as new capabilities are constantly being added to the framework and the module template is updated to reflect them.

**Comment by vnetonline** [comment]
@sbwalker I have used the Create Module component included in the Oqtane Framework, which leverages the Default Template included with Oqtane below, you can see the Create Module Solution Screen  

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/28659643-20ae-483d-8164-a7eec4cf69c6)

- I have added some lines of code to `CreateModule` Method to write `solution.Json` into `.amazing.oqtane.studio` folder at the root of the module solution

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/16a987dc-cf21-47be-8601-1cd168d62a23)

- The solution.json includes the version number

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/e369ce56-91de-4efc-a08e-48cc332c730b)

Amazing Oqtane Studio allows you to build Oqtane Modules in a matter of minutes. The Amazing thing is that you don't have to worry about building repetitive code. It is specially built for the Oqtane Framework as a Oqtane module and it leverages the Default Module Template

https://learnoqtane.com/modules/amazing-oqtane-studio

Amazing Oqtane Studio generates all the necessary CRUD code for you in a few seconds following the Oqtane Framework conventions.

Since you have the source code, you can easily modify the source code and inject your custom/business logic into the generated code.

Amazing oqtane Studio generates the source code for the below-mentioned aspects of your module solution.

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/5f902828-d3ed-46dd-959a-b18911f8a9f1)

**Comment by vnetonline** [comment]
@sbwalker or @leigh-pointer  I am up to generating the code for `[Module]DBContext` and `01000000_InitalizeModule`.

My question should Amazing.Oqtane.Studio generate a separate `[Entity]DBContext` and a  `01000000_Initalize[Entity]` for each entity generated by Amazing.Oqtane.Studio?

Or The generated code has to be inserted in  `[Module]DBContext` and  `01000000_InitalizeModule`.

**Reply by leigh-pointer** [reply]
@vnetonline i might be a inclined to have a yes/no option to say which, separate or single!

**Reply by vnetonline** [reply]
@leigh-pointer  by default what option would be best?

**Reply by leigh-pointer** [reply]
Default i think it should be separate as this keeps the Modules module.

**Reply by vnetonline** [reply]
So, say if there are two entities Doctor and Patient then I should generate. 

PatientContext and 01000000_InitiazePatient
DoctorContext and 01000000_InitializeDoctor

**Comment by vnetonline** [comment]
Amazing Oqtane Studio is now able to generate code and add it to your  [Module]Context and Initalize[Module] for your Entities and EntityBuilders

![screencast-amazing-oqtane-studio-InitializeModule-ModuleContext](https://github.com/oqtane/oqtane.framework/assets/15242923/3d974ea8-3f4b-4d13-ac13-044ae6efb04c)

**Comment by vnetonline** [comment]
I have created a repo to track Issues and Discussions about [Amazing Oqtane Studio](https://github.com/AmazingSoftwareSolutions/Amazing.Oqtane.Studio/)

Since this commercial product will be sold at Oqtane Market Place, the code has not been uploaded to this repo. 

This repo has been created to include support documents, issues, and discussions for Amazing Oqtane Studio. 

If you want to add enhancements or feature requests, please raise an Issue in the repo.

Learn more about Amazing Oqtane Studio by following the link below 

https://github.com/AmazingSoftwareSolutions/Amazing.Oqtane.Studio/

**Reply by vnetonline** [reply]
I have decided to provide the Amazing Oqtane Studio as a SaaS service register your interest at https://amazingoqtanestudio.com

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/80295b99-b4f6-4345-a69a-5649cb6b6239)

**Comment by vnetonline** [comment]
Amazing Oqtane Studio can now generate [Enity]Repository for each model

Check it out by following the  below link 

https://github.com/AmazngSoftwareSolutions/Amazing.Oqtane.Studio/issues/8#issue-2301709346

**Comment by vnetonline** [comment]
Amazing Oqtane Studio can now generate [Entity]Manager for each model

https://github.com/AmazingSoftwareSolutions/Amazing.Oqtane.Studio/issues/9#issue-2301781565

**Comment by vnetonline** [comment]
I have decided to provide the Amazing Oqtane Studio as a SaaS service. Register your interest at https://amazingoqtanestudio.com

![image](https://github.com/oqtane/oqtane.framework/assets/15242923/35ccebaa-b846-4dd9-9975-75f57f80fc2d)

The first 5 Registrations will get a free subscription.

**Comment by vnetonline** [comment]
Amazing Oqtane Studio (AOS) has been launched today. 

FREE Membership Now Available For Limited Time Build Oqtane Modules Fast by igniting Oqtane (rocket fuel) for Blazor!

https://amazingoqtanestudio.com
